File: oldbackend/edit_distance.py
from feedback_types import Range
from typing import Any

class EditOp:
    def __init__(self, range:Range, newString:str):
        self.range = range
        self.newString = newString

    def __repr__(self):
        return f"EditOperation(Range({self.range.start}, {self.range.end}), '{self.newString}')"

    def to_json(self) -> dict[str,Any]:
        return {
            "range": self.range.to_json(),
            "newString": self.newString
        }


def edit_distance_operations_with_classes(str1:str, str2:str):

    # Create a matrix for storing edit distances
    m, n = len(str1), len(str2)
    dp = [[0] * (n + 1) for _ in range(m + 1)]

    # Initialize the first row and column
    for i in range(m + 1):
        dp[i][0] = i  # Deletion cost
    for j in range(n + 1):
        dp[0][j] = j  # Insertion cost

    # Compute the edit distance matrix
    for i in range(1, m + 1):
        for j in range(1, n + 1):
            if str1[i - 1] == str2[j - 1]:
                dp[i][j] = dp[i - 1][j - 1]  # No operation needed
            else:
                dp[i][j] = 1 + min(dp[i - 1][j],      # Delete
                                   dp[i][j - 1],      # Insert
                                   dp[i - 1][j - 1])  # Replace

    # Trace back to find the operations
    operations = []
    i, j = m, n
    while i > 0 and j > 0:
        if str1[i - 1] == str2[j - 1]:
            i, j = i - 1, j - 1
        elif dp[i][j] == dp[i - 1][j - 1] + 1:  # Replace
            operations.append({'range': Range(i - 1, i), 'newString': str2[j - 1]})
            i, j = i - 1, j - 1
        elif dp[i][j] == dp[i - 1][j] + 1:  # Delete
            operations.append({'range': Range(i - 1, i), 'newString': ''})
            i -= 1
        elif dp[i][j] == dp[i][j - 1] + 1:  # Insert
            operations.append({'range': Range(j - 1, j), 'newString': str2[j - 1]})
            j -= 1

    # Handle remaining characters in str1 (deletions) or str2 (insertions)
    while i > 0:
        operations.append({'range': Range(i - 1, i), 'newString': ''})
        i -= 1
    while j > 0:
        operations.append({'range': Range(j - 1, j), 'newString': str2[j - 1]})
        j -= 1



    # Consolidate adjacent operations
    consolidated_operations:list[EditOp] = []
    for op in operations[::-1]:
        if consolidated_operations and consolidated_operations[-1].range.is_adjacent(op['range']):
            consolidated_operations[-1].range.merge(op['range'])
            consolidated_operations[-1].newString += op['newString']
        else:
            consolidated_operations.append(EditOp(op['range'], op['newString']))

    return consolidated_operations

# Example usage
# print(edit_distance_operations_with_classes("kitten", "sittinggg"))

File: frontend/src/api/editDistance.ts
import { DocRange } from "@api/ApiTypes";

class EditOp {
    range: DocRange;
    newString: string;

    constructor(range: DocRange, newString: string) {
        this.range = range;
        this.newString = newString;
    }

    toString() {
        return `EditOperation(Range(${this.range.start}, ${this.range.end}), '${this.newString}')`;
    }
}

export function editDistanceOperationsWithClasses(
    str1: string,
    str2: string,
): EditOp[] {
    // Create a matrix for storing edit distances
    const m = str1.length,
        n = str2.length;
    const dp: number[][] = Array(m + 1)
        .fill(0)
        .map(() => Array(n + 1).fill(0));

    // Initialize the first row and column
    for (let i = 0; i <= m; i++) {
        dp[i][0] = i; // Deletion cost
    }
    for (let j = 0; j <= n; j++) {
        dp[0][j] = j; // Insertion cost
    }

    // Compute the edit distance matrix
    for (let i = 1; i <= m; i++) {
        for (let j = 1; j <= n; j++) {
            if (str1[i - 1] == str2[j - 1]) {
                dp[i][j] = dp[i - 1][j - 1]; // No operation needed
            } else {
                dp[i][j] =
                    1 +
                    Math.min(
                        dp[i - 1][j], // Delete
                        dp[i][j - 1], // Insert
                        dp[i - 1][j - 1],
                    ); // Replace
            }
        }
    }

    // Trace back to find the operations
    const operations = [];
    let i = m,
        j = n;
    while (i > 0 && j > 0) {
        if (str1[i - 1] == str2[j - 1]) {
            i--;
            j--;
        } else if (dp[i][j] == dp[i - 1][j - 1] + 1) {
            // Replace
            operations.push(new EditOp(new DocRange(i - 1, i), str2[j - 1]));
            i--;
            j--;
        } else if (dp[i][j] == dp[i - 1][j] + 1) {
            // Delete
            operations.push(new EditOp(new DocRange(i - 1, i), ""));
            i--;
        } else if (dp[i][j] == dp[i][j - 1] + 1) {
            // Insert
            operations.push(new EditOp(new DocRange(j - 1, j), str2[j - 1]));
            j--;
        }
    }

    // Handle remaining characters in str1 (deletions) or str2 (insertions)
    while (i > 0) {
        operations.push(new EditOp(new DocRange(i - 1, i), ""));
        i--;
    }
    while (j > 0) {
        operations.push(new EditOp(new DocRange(j - 1, j), str2[j - 1]));
        j--;
    }

    // Consolidate adjacent operations
    const consolidated_operations: EditOp[] = [];
    for (const op of operations.reverse()) {
        if (
            consolidated_operations.length &&
            consolidated_operations[
                consolidated_operations.length - 1
            ].range.isAdjacent(op.range)
        ) {
            consolidated_operations[
                consolidated_operations.length - 1
            ].range.merge(op.range);
            consolidated_operations[
                consolidated_operations.length - 1
            ].newString += op.newString;
        } else {
            consolidated_operations.push(new EditOp(op.range, op.newString));
        }
    }

    return operations;
}

// Example usage
// console.log(editDistanceOperationsWithClasses("kitten", "sittinggg"));
